src/python/pants/base/exception_sink.py

```diff
@@ -19,28 +24,181 @@
 class ExceptionSink(object):
   """A mutable singleton object representing where exceptions should be logged to."""
 
-  _destination = os.getcwd()
+  # NB: see the bottom of this file where we call reset_log_location() and other mutators in order
+  # to properly setup global state.
+  _log_dir = None
+  # We need an exiter in order to know what to do after we log a fatal exception.
+  _exiter = None
+  # Where to log stacktraces to in a SIGUSR2 handler.
+  _interactive_output_stream = None
+
+  # These persistent open file descriptors are kept so faulthandler can figure out signal safety.
+  _pid_specific_error_fileobj = None
+  _shared_error_fileobj = None
+
+  # Integer code to exit with on an unhandled exception.
+  UNHANDLED_EXCEPTION_EXIT_CODE = 1
 
   def __new__(cls, *args, **kwargs):
     raise TypeError('Instances of {} are not allowed to be constructed!'
                     .format(cls.__name__))
 
+  class ExceptionSinkError(Exception): pass
+
+  # All reset_* methods are ~idempotent!
+  @classmethod
+  def reset_log_location(cls, new_log_location=None):
+    """
+    Class state:
+    - Overwrites `cls._log_dir`, `cls._pid_specific_error_fileobj`, and
+      `cls._shared_error_fileobj`.
+    OS state:
+    - May create a new directory.
+    - Overwrites signal handlers for many fatal signals.
+
+    :raises: :class:`ExceptionSink.ExceptionSinkError` if the directory is not writable.
+    """
+    # We could no-op here if the log locations are the same, but there's no reason not to have the
+    # additional safety of re-acquiring file descriptors each time.
+
+    if new_log_location:
+      # Create the directory if possible, or raise if not writable.
+      cls._check_or_create_new_destination(new_log_location)
+    else:
+      # Interpret a None argument to mean to just refresh the file handles for the current pid.
```

Comment:
This method is never called with `None`, afaict, so would be good to remove this case to improve clarity.

Reply:
Yes, after talking with you last week about where we need to reset the logs (whenever we get bootstrap options or whenever we change pid) and we can remove that now.